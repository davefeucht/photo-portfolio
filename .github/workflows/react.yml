# This workflow will do a clean install of node dependencies, build the source code and run tests across different versions of node
# For more information see: https://help.github.com/actions/language-and-framework-guides/using-nodejs-with-github-actions

name: React CI/CD

on:
  push:
    branches: [ master ]
  pull_request:
    branches: [ master ]

jobs:
  build:

    runs-on: ubuntu-latest

    strategy:
      matrix:
        node-version: [18.x]

    steps:
    - name: Checkout code
      uses: actions/checkout@v2

    - name: Install Node.js ${{ matrix.node-version }}
      uses: actions/setup-node@v1
      with:
        node-version: ${{ matrix.node-version }}

    - name: Enable corepack
      run: corepack enable

    - name: Install dependencies
      run: yarn install

    - name: Build project
      run: npm run build --if-present

    - name: Run tests
      run: npm test
