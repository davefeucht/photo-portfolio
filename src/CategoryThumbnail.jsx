/*****************
* CategoryThumbnail component displays the post image for a particular category.
* Props:
*  id: PropTypes.number,
*  name: PropTypes.string,
*  site: PropTypes.string,
*  showAllPosts: PropTypes.bool,
*  clickCategory: PropTypes.func
*****************/

import React from "react";
import axios from "axios";
import PropTypes from "prop-types";

export default class CategoryThumbnail extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      categoryClasses: ["category"],
      categoryPost: {},
      errorMsg: ""
    };
  }

  //Function to get the 'main' post for the category
  _getCategoryPost(categoryId) {
    let getCategoryPostURI = "http://" + this.props.site + "/wp-json/wp/v2/posts?categories=" + categoryId;
    axios.get(getCategoryPostURI)
      .then(res => {
        if(res.data[0] !== undefined) {
          const categoryPost = res.data[0];
          this.setState({categoryPost});
        }
      })
      .catch(error => {
        const errorMsg = "Could not get posts for this category: " + (error.response ? error.response : error);
        this.setState({ errorMsg });
      });
  }

  _highlightCategory() {
    if(!this.state.categoryClasses.includes("hovered")) {
      const categoryClasses = this.state.categoryClasses.concat(["hovered"]);
      this.setState({categoryClasses});
    }
  }

  _unHighlightCategory() {
    if(this.state.categoryClasses.includes("hovered")) {
      const categoryClasses = this.state.categoryClasses.filter(element => element !== "hovered");
      this.setState({categoryClasses});
    }
  }

  //When the component is about to mount, get the main post for the category
  componentWillMount() {
    this._getCategoryPost(this.props.id);
  }

  render() {
    const divStyle = {backgroundImage: "url(" + this.state.fullImageUrl + ")", backgroundRepeat: "no-repeat", backgroundPosition: "center center", backgroundSize: "auto 100%"};

    return ( 
      <div style={divStyle} className={this.state.categoryClasses.join(" ")} onClick={this._openCategory.bind(this)} onMouseOver={this._highlightCategory.bind(this)} onMouseOut={this._unHighlightCategory.bind(this)}>
      </div>
    );
  }

}

CategoryThumbnail.propTypes = {
  id: PropTypes.number,
  name: PropTypes.string,
  site: PropTypes.string,
  showAllPosts: PropTypes.bool,
  clickCategory: PropTypes.func
};