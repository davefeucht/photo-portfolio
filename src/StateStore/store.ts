import { configure, makeAutoObservable } from 'mobx';
import { ImageData, Image, Store, Category, Post, Page } from '../utils/types';

configure({ enforceActions: 'observed' });

class stateStore implements Store {
    menuState = 'closed'
    applicationRoot: HTMLElement = null;
    screenInfo: {
        width: number,
        height: number
    }
    siteInfo: {
        siteName: string,
        siteUrl: string
    }
    visiblePost: {
        postId: number,
        postTitle: string,
        tags: number[],
        tagNames: string[],
        fullImageUrl: string,
        width: number,
        height: number
    };
    categoryList: Category[] = [];
    currentCategoryPosts: Post[] = [];
    currentCategoryData: Category = {
        id: 0,
        count: 0,
        description: '',
        link: '',
        name: '',
        slug: '',
        taxonomy: '',
        parent: 0
    };
    pages: Page[] = [];
    currentPageData: Page = {
        id: 0,
        link: '',
        modified: '',
        slug: '',
        type: '',
        title: {
            rendered: ''
        },
        content: {
            rendered: '',
            protected: false
        },
        author: 0,
        featured_media: 0
    };
    maxItemsPerPage = 10;

    constructor() {
        makeAutoObservable(this);
    }

    setMenuState = (state: string) => {
        this.menuState = state;
    }

    toggleMenuState = () => {
        this.menuState = this.menuState === 'closed' ? 'open' : 'closed';
    }

    setApplicationRoot = (element: HTMLElement) => {
        this.applicationRoot = element;
    }

    setSiteName = (name: string) => {
        this.siteInfo.siteName = name;
    }

    setVisiblePost = (postId: number, postTitle: string) => {
        this.visiblePost.postId = postId;
        this.visiblePost.postTitle = postTitle;
    }

    setVisiblePostImage = (fullImageUrl: string) => {
        this.visiblePost.fullImageUrl = fullImageUrl;
    }

    setVisiblePostTags = (tagNames: string[]) => {
        tagNames.forEach(tagName => {
            this.visiblePost.tagNames.push(tagName);
        });
    }

    clearVisiblePostTagNames = () => {
        this.visiblePost.tagNames.splice(0, this.visiblePost.tagNames.length);
    }

    setCategoryList = (categories: Category[]) => {
        this.categoryList.length = 0;
        categories.forEach((category, index) => {
            this.categoryList[index] = category;
        })
    }

    setCategoryPosts = (posts: Post[]) => {
        this.currentCategoryPosts.length = 0;
        posts.forEach((post, index) => {
            this.currentCategoryPosts[index] = post;
        })
    }

    setCategoryData = (categoryData: Category) => {
        Object.keys(categoryData).forEach((key: keyof Category) => {
            this.currentCategoryData[key] = categoryData[key];
        });
    }

    setThumbnailImageUrl = (imageData: ImageData) => {
        this.currentCategoryPosts[imageData.post_index].thumbnail_image = imageData.image_url;
    }

    setCurrentPost = (postData: Post) => {
        Object.keys(postData).forEach((key: keyof Image) => {
            this.visiblePost[key] = postData[key];
        })
    }

    setPages = (pages: Page[]) => {
        pages.forEach(page => {
            this.pages.push(page);
        });
    }

    setPageData = (pageData: Page) => {
        Object.assign(this.currentPageData, pageData);
    }
}

export default stateStore;
