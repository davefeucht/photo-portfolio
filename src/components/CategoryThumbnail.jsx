/*****************
* CategoryThumbnail component displays the post image for a particular category.
*****************/

import React from "react";
import axios from "axios";
import PropTypes from "prop-types";
import StyledCategoryThumbnail from "./styledComponents/StyledCategoryThumbnail.jsx";
import {connect} from "react-redux";
import {bindActionCreators} from "redux";
import {setCategoryList, setCategoryThumbnail} from "../actions/actions.js";

class CategoryThumbnail extends React.Component {

  static propTypes = {
    id: PropTypes.number,
    index: PropTypes.number,
    name: PropTypes.string,
    siteUrl: PropTypes.string,
    clickCategory: PropTypes.func
  };

  //Function to get a random post image URL for the category
  _getCategoryImage(categoryId) {
    //Define the request string to get the posts for this category
    const getCategoryPostURI = `https://${this.props.siteUrl}/wp-json/wp/v2/posts?categories=${categoryId}`;
    //Make the request
    axios.get(getCategoryPostURI) 
      .then((response) => {
        //Determine a random post from the ones returned
        const numberOfPosts = response.data.length;
        const randomPost = Math.floor((Math.random() * (numberOfPosts + 1)));

        //If this post does exist in the returned results
        if(response.data[randomPost] !== undefined) {
          //Define the request string to get the featured media for the random post
          const getCategoryImage = `https://${this.props.siteUrl}/wp-json/wp/v2/media/${response.data[randomPost].featured_media}/`; 
          //Make the request
          axios.get(getCategoryImage)
            .then(response => {
              let fullImageUrl = "";

              //If the large size image exists, use it
              if(response.data.media_details.sizes.large) {
                fullImageUrl = response.data.media_details.sizes.large.source_url;
              }
              //Otherwise use the medium size version
              else {
                fullImageUrl = response.data.media_details.sizes.medium.source_url;
              }
              this.props.setCategoryThumbnail({category_index: this.props.index, image_url: fullImageUrl});
            })
        }
      })
  }

  //Function to open the category using the function passed in from the parent component
  _openCategory() {
    this.props.clickCategory(this.props.id, this.props.name);
  }

  _highlightCategory() {
    /*
    if(!this.state.categoryClasses.includes("hovered")) {
      const categoryClasses = this.state.categoryClasses.concat(["hovered"]);
      this.setState({categoryClasses});
    }
    */
  }

  _unHighlightCategory() {
    /*
    if(this.state.categoryClasses.includes("hovered")) {
      const categoryClasses = this.state.categoryClasses.filter(element => element !== "hovered");
      this.setState({categoryClasses});
    }
    */
  }

  //When the component is about to mount, get the random post for the category
  componentWillMount() {
    this._getCategoryImage(this.props.id);
  }

  render() {
    const divStyle = {backgroundImage: "url(" + (this.props.categoryList[this.props.index].thumbnail_image ? this.props.categoryList[this.props.index].thumbnail_image : "")+ ")", backgroundRepeat: "no-repeat", backgroundPosition: "center center", backgroundSize: "auto 100%"};

    return ( 
      <StyledCategoryThumbnail style={divStyle} className="category-thumbnail" /*{this.state.categoryClasses.join(" ")}*/ onClick={this._openCategory.bind(this)} onMouseOver={this._highlightCategory.bind(this)} onMouseOut={this._unHighlightCategory.bind(this)}>
      </StyledCategoryThumbnail>
    );
  }
}

const mapStateToProps = state => {
  return {
    siteUrl: state.applicationState.siteUrl,
    categoryList: state.applicationState.categoryList
  };
}

const mapDispatchToProps = dispatch => {
  return {
    setCategoryThumbnail: bindActionCreators(setCategoryThumbnail, dispatch)
  }; 
}

export default connect(mapStateToProps, mapDispatchToProps)(CategoryThumbnail);