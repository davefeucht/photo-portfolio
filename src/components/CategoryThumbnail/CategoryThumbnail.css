.category-thumbnail {
  display: flex;
  flex-grow: 1;
  cursor: pointer;
  background-repeat: no-repeat;
  background-position: center center;
  background-size: cover;
}