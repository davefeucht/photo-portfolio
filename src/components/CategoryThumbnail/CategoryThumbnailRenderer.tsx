import ImageListItem from "@mui/material/ImageListItem";
import ImageListItemBar from '@mui/material/ImageListItemBar';
import * as React from 'react';
import { Link } from 'react-router-dom';

interface Props {
    id: number;
    name: string;
    thumbnailUrl: string;
}

const CategoryThumbnailRenderer: React.FC<Props> = ({
    id,
    name,
    thumbnailUrl
}) => {
    return (
        <Link to={`/category/${id}`}>
            {thumbnailUrl && (
                <ImageListItem sx={{ overflow: "hidden" }}>
                    <img
                        src={thumbnailUrl}
                        alt={name}
                        loading="lazy"
                    />
                    <ImageListItemBar
                        title={name}
                    />
                </ImageListItem>
            )}
        </Link>
    );
};

export default CategoryThumbnailRenderer;
