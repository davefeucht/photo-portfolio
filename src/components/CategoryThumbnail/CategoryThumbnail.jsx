/*****************
* CategoryThumbnail component displays the post image for a particular category.
*****************/

import React from 'react';
import PropTypes from 'prop-types';
import { observer } from 'mobx-react';
import { Link } from 'react-router-dom';
import CategoryTitle from '../CategoryTitle/CategoryTitle.jsx';

import { getCategoryImage } from '../../utils/Api';

import './CategoryThumbnail.css';

const CategoryThumbnail = ({ id, index, name, stateStore }) => {

    if (!stateStore.categoryList[index].thumbnail_image) {
        getCategoryImage(id, index, stateStore);
    }

    const divStyle = { backgroundImage: "url(" + (stateStore.categoryList[index].thumbnail_image ? stateStore.categoryList[index].thumbnail_image : "") + ")" };

    return (
        <Link to={`/category/${id}`}>
            <div style={divStyle} className="category-thumbnail">
                <CategoryTitle title={name} />
            </div>
        </Link>
    );
};

CategoryThumbnail.displayName = 'CategoryThumbnail';

CategoryThumbnail.propTypes = {
    id: PropTypes.number.isRequired,
    index: PropTypes.number.isRequired,
    name: PropTypes.string.isRequired,
    stateStore: PropTypes.object.isRequired
};

export default observer(CategoryThumbnail);
