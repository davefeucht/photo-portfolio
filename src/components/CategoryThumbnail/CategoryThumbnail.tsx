/** ***************
* CategoryThumbnail component displays the post image for a particular category.
**************** */

import './CategoryThumbnail.css';

import { observer } from 'mobx-react';
import * as React from 'react';
import { useState } from 'react';
import { Link } from 'react-router-dom';

import { getCategoryImage } from '../../utils/Api';
import { Store } from '../../utils/types';
import CategoryTitle from '../CategoryTitle/CategoryTitle';

interface CategoryThumbnailProps {
    id: number,
    name: string,
    stateStore: Store
}

const CategoryThumbnail: React.FC<CategoryThumbnailProps> = ({ id, name, stateStore }) => {
    const [thumbnailImageUrl, setThumbnailImageUrl] = useState('');

    if (!thumbnailImageUrl) {
        getCategoryImage(id, stateStore.siteInfo.siteUrl)
            .then(imageUrl => {
                setThumbnailImageUrl(imageUrl);
            });
    }

    const divStyle = { backgroundImage: `url(${thumbnailImageUrl || ''})` };

    return (
        <Link to={`/category/${id}`}>
            <div style={divStyle} className="category-thumbnail">
                <CategoryTitle title={name} />
            </div>
        </Link>
    );
};

CategoryThumbnail.displayName = 'CategoryThumbnail';

export default observer(CategoryThumbnail);
