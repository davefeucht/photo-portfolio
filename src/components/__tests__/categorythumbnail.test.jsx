import { render } from '@testing-library/react';
import React from 'react';
import { MemoryRouter } from 'react-router-dom';

import StateStore from '../../StateStore/store';
import CategoryThumbnail from '../CategoryThumbnail/CategoryThumbnail';

const id = 5;
const name = 'Test Category';

jest.mock('../../utils/Api', () => ({
    getCategoryImage: () => Promise.resolve('https://throughapinhole.com/wp-content/uploads/2018/07/DSC_1508.jpg')
}));

test('CategoryThumbnail displays', async () => {
    const store = new StateStore();
    const { container } = render(
        <MemoryRouter>
            <CategoryThumbnail id={id} name={name} stateStore={store} />
        </MemoryRouter>
    );
    expect(container.firstChild).toMatchSnapshot();
});
