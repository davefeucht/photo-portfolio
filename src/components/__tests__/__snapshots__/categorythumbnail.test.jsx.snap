// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CategoryThumbnail displays 1`] = `
<a
  href="/category/5"
>
  <div
    class="category-thumbnail"
    style="background-image: url();"
  >
    <div
      class="category-title"
    >
      <div
        class="category-title-text"
      >
        Test Category
      </div>
    </div>
  </div>
</a>
`;
